Opentelemetry transfer choice

Deexie opened the pull request:
Having been fixed after review, the PR now contains 3 commits as follows:
- tracing: add OpenTelementry trace state to trace_state_ptr - puts information about opentelemetry tracing in trace_state_ptr. 
New: OpenTelemetry tracing flag is set in constructors of opentelemetry_state.
- instantiate trace_state_ptr basing on user's tracing choice - basing on the result of negotiation with driver proper fields of trace_state_ptr are set.
- pass content of opentelemetry_state to response - otel tracing info is sent to client iff OpenTelemetry tracing was requested.
New: instead of {"abc", "abc"} a map with serialized tracing data is passed (empty yet though); make_unique_response was moved to tracing/response.hh.

All previous PRs were updated too so partial review/merge could be done.

margdoc committed:
tracing: add OpenTelementry trace state to trace_state_ptr

Add opentelemetry_state pointer to trace_state_ptr.
Move trace_state pointer to opentelemetry_state
for best-case scenario optimisation.

wprzytula reviewed (changes requested):
(unintended)

constantfold commented on cql3/query_processor.hh:
```diff
@@ -123,6 +123,7 @@ private:
     service::migration_notifier& _mnotifier;
     service::migration_manager& _mm;
     const cql_config& _cql_config;
+    bool _tracing_required = false;
```
Unfortunately, this approach with `_tracing_required` on `query_processor` is not correct. For example, it will be only set if the client ever sends a `PREPARE` request. There is only one `query_processor` per `transport/server.cc::server`, so `_tracing_required` would be incorrectly reused between different sessions.

The approach you should take is to reuse the machinery behind the "classic" tracing. There it works this way: if the user requested "classic" tracing on a request, it builds non-empty `trace_state_ptr`. Otherwise, an "empty" `trace_state_ptr` is passed around.

Here it figures out what type of tracing should be used: https://github.com/scylladb/scylla/blob/6e2d29300c79a8f5af76641972e5be7d9d381bf9/transport/server.cc#L601-L608

Here it actually builds a `trace_state_ptr`: https://github.com/scylladb/scylla/blob/6e2d29300c79a8f5af76641972e5be7d9d381bf9/transport/server.cc#L388

Your code should correctly fill out `trace_state_ptr` based on whether the client has `OPENTELEMETRY_TRACING`.

constantfold commented on tracing/trace_state.hh:
```diff
@@ -505,28 +505,113 @@ private:
+class opentelemetry_state final {
 private:
     lw_shared_ptr<trace_state> _state_ptr;
+    bool const _opentelemetry_tracing{false};
```
Per my previous comment, it should be dynamically filled in (not const!)

margdoc replied:
I think making it const makes it similar to "classic" tracing. As you said "There it works this way: if the user requested "classic" tracing on a request, it builds non-empty trace_state_ptr. Otherwise, an "empty" trace_state_ptr is passed around.". The decision if opentelemetry is used is made somewhere at the beginning of request processing.

constantfold replied:
Yes, what I meant is that the problem is that the `_openetelemetry_tracing` can never be `true` - hardcoded to `false` and not properly filled in.

Deexie committed:
instantiate trace_state_ptr basing on user's tracing choice

Depending on the type of tracing (classic, opentelemetry, both
or none) chosen by client, trace_state_ptr is properly instantiated.

jul-stas commented on transport/response.hh:
```diff
@@ -141,6 +141,16 @@ private:
     }
 };
 
+inline std::unique_ptr<cql_server::response> make_unique_response(int16_t stream, cql_binary_opcode opcode, const tracing::trace_state_ptr& tr_state_ptr) {
```
After some discussion we believe that the cleanest solution will be to set up the custom payload in the ctor of `response`. So we'd have one ctor: `response(int16_t, cql_binary_opcode, const tracing::trace_state_ptr&)` and we'd move the logic of this function into it.
That way we don't edit any unrelated code in `server.cc`, potentially avoiding merge conflicts in the future.

constantfold commented on tracing/tracing.cc:
```diff
@@ -130,7 +130,7 @@ trace_state_ptr tracing::create_session(trace_type type, trace_state_props_set p
 
         ++_active_sessions;
-        return make_lw_shared<trace_state>(type, props);
+        return make_lw_shared<opentelemetry_state>(make_lw_shared<trace_state>(type, props), opentelemetry_tracing);
```
In line `125` of this file (cannot add a comment at that site), there is this `if`:

```cpp
if (!may_create_new_session()) {
    return nullptr;
}
```

which will cause OpenTelemetry tracing to be erroneously disabled.

constantfold commented on tracing/tracing.cc:
```diff
@@ -115,7 +115,7 @@ future<> tracing::stop_tracing() {
-trace_state_ptr tracing::create_session(trace_type type, trace_state_props_set props) noexcept {
+trace_state_ptr tracing::create_session(trace_type type, trace_state_props_set props, bool opentelemetry_tracing) noexcept {
     if (!started()) {
         return nullptr;
```
Similar problem here (OpenTelemetry tracing erroneously disabled)

constantfold commented on tracing/trace_state.hh:
```diff
@@ -505,28 +505,118 @@ private:
+     * @return True if classic trace state is stored.
+     */
+    bool has_tracing() const noexcept {
+        return bool(_state_ptr);
```
```suggestion
        return __builtin_expect(bool(_state_ptr), false);
```

constantfold commented on tracing/trace_state.hh:
```diff
@@ -505,28 +505,118 @@ private:
+     * @return True if classic trace state is stored.
+     */
+    bool has_tracing() const noexcept {
+        return has_any_tracing() && _state_ptr->has_tracing();
```
```suggestion
        return __builtin_expect(has_any_tracing() && _state_ptr->has_tracing(), false);
```

constantfold commented on tracing/trace_state.hh:
```diff
@@ -505,28 +505,118 @@ private:
+     * @return True if OpenTelemetry trace state is stored.
+     */
+    bool has_opentelemetry() const noexcept {
+        return has_any_tracing() && _state_ptr->has_opentelemetry();
```
```suggestion
        return __builtin_expect(has_any_tracing() && _state_ptr->has_opentelemetry(), false);
```

although the end goal is to have OpenTelemetry turned on all the time, this (should?) minimize the overhead when it's not turned on.

avelanarius reviewed (approved):
LGTM, apart from the comments I left.

Deexie committed:
pass content of opentelemetry_state to response

If OpenTelemetry tracing was requsted content of opentelemetry_state
is serialized and passed to client in custom payload of response.
